src/mumble/GlobalShortcut_win.cpp

```diff
@@ -719,211 +421,170 @@ void GlobalShortcutWin::timeTicked() {
+
 					handleButton(ql, pressed);
 				}
 
-				xboxinputLastPacket[i] = state.packetNumber;
+				m_xinputLastPacket[i] = state.packetNumber;
 			}
 		}
 	}
 #endif
+#ifdef USE_GKEY
+	if (m_gkey) {
+		for (uint8_t btn = GKEY_MIN_MOUSE_BUTTON; btn <= GKEY_MAX_MOUSE_BUTTON; ++btn) {
+			const QList< QVariant > ql{ QVariant::fromValue(DeviceTypes::GKEY_MOUSE), btn };
 
-	// Initialize winhooks.
-	//
-	// We do this here, because at this point, we've just run our
-	// first timeTicked() slot. The GlobalShortcut_win thread's event
-	// loop has nothing else to do at this point, so there is nothing
-	// that blocks the callbacks of the hooks.
-	//
-	// That is, if we initialize here, our callbacks *can* be called
-	// immediately after initialization, which gives the best results
-	// as far as interactivity and user experience goes. The initialization
-	// cannot be "felt".
-	//
-	// Let me explain...
-	//
-	// Originally, this code lived in the body of run, ::run(), just
-	// before exec() was called.
-	//
-	// It turns out that if our hooks are initialized there, it can take
-	// a short while before the mouse and keyboard callbacks can be processed.
-	//
-	// During this time, where the mouse callback is not able to be called,
-	// the mouse in Windows becomes laggy. It makes the whole computer feel
-	// like it has locked up -- because input "stops".
-	//
-	// As explained above, initializing the hooks here yields a much superior
-	// experience, where this initialization has no observable effect on the
-	// behavior of the system's mouse input.
-	if (bHook && !hhMouse && !hhKeyboard) {
-		HMODULE hSelf;
-		GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
-						  reinterpret_cast< LPCTSTR >(&HookKeyboard), &hSelf);
-		hhMouse    = SetWindowsHookEx(WH_MOUSE_LL, HookMouse, hSelf, 0);
-		hhKeyboard = SetWindowsHookEx(WH_KEYBOARD_LL, HookKeyboard, hSelf, 0);
-	}
-}
+			handleButton(ql, m_gkey->isMouseButtonPressed(btn));
+		}
 
-GlobalShortcutWin::ButtonInfo GlobalShortcutWin::buttonInfo(const QVariant &v) {
-	GlobalShortcutWin *gsw = static_cast< GlobalShortcutWin * >(GlobalShortcutEngine::engine);
+		for (uint8_t mode = GKEY_MIN_KEYBOARD_MODE; mode <= GKEY_MAX_KEYBOARD_MODE; mode++) {
+			for (uint8_t key = GKEY_MIN_KEYBOARD_BUTTON; key <= GKEY_MAX_KEYBOARD_BUTTON; key++) {
+				const QList< QVariant > ql{ QVariant::fromValue(DeviceTypes::GKEY_KEYBOARD),
+											// Store the key and mode in one int
+											// bit 0..15: mode, bit 16..31: key
+											key | (mode << 16) };
 
-	const QList< QVariant > &sublist = v.toList();
-	if (sublist.count() != 2) {
-		return ButtonInfo();
+				handleButton(ql, m_gkey->isKeyboardGkeyPressed(key, mode));
+			}
+		}
 	}
+#endif
+}
 
-	bool ok    = false;
-	DWORD type = sublist.at(0).toUInt(&ok);
-	QUuid guid(sublist.at(1).toString());
+GlobalShortcutWin::ButtonInfo GlobalShortcutWin::buttonInfo(const QVariant &button) {
```

Comment:
I wonder if the code might get a bit more elegant if we used a custom struct with named fields instead of a variant that implicitly encapsulates a list that has certain elements at certain positions :thinking:

Reply:
We would still have to use a `QVariant` though, because each input backend has its own structure.

Reply:
We _could_ also use Polymorphism, but I guess it is not worth it to deal with ownership issues for data passed through Qt's hands.

I think I would consider the variant of different structs cleaner than the variants of different lists though.

Reply:
A list would still be required though, to know the structure type.

Reply:
Afaik we can query the `QVariant` for its type. Thus I don't think we need a list for that

Reply:
You're right, we can get the meta type ID.